soroban-sdk/src/events.rs

```diff
@@ -0,0 +1,131 @@
+use core::fmt::Debug;
+
+#[cfg(doc)]
+use crate::{contracttype, Bytes, BytesN, Map};
+use crate::{env::internal, vec, Env, IntoVal, RawVal, Vec};
+
+// TODO: consolidate with host::events::TOPIC_BYTES_LENGTH_LIMIT
+const TOPIC_BYTES_LENGTH_LIMIT: u32 = 32;
+
+/// Events publishes events for the currently executing contract.
+///
+/// ```
+/// use soroban_sdk::Env;
+///
+/// # use soroban_sdk::{contractimpl, vec, map, RawVal, BytesN};
+/// #
+/// # pub struct Contract;
+/// #
+/// # #[contractimpl]
+/// # impl Contract {
+/// #     pub fn f(env: Env) {
+/// let event = env.events();
+/// let data = map![&env, (1u32, 2u32)];
+/// let topics0 = ();
+/// let topics1 = (0u32,);
+/// let topics2 = (0u32, 1u32);
+/// let topics3 = (0u32, 1u32, 2u32);
+/// let topics4 = (0u32, 1u32, 2u32, 3u32);
+/// event.publish(topics0, data.clone());
+/// event.publish(topics1, data.clone());
+/// event.publish(topics2, data.clone());
+/// event.publish(topics3, data.clone());
+/// event.publish(topics4, data.clone());
+/// #     }
+/// # }
+///
+/// # #[cfg(feature = "testutils")]
+/// # fn main() {
+/// #     let env = Env::default();
+/// #     let contract_id = BytesN::from_array(&env, &[0; 32]);
+/// #     env.register_contract(&contract_id, Contract);
+/// #     f::invoke(&env, &contract_id);
+/// # }
+/// # #[cfg(not(feature = "testutils"))]
+/// # fn main() { }
+/// ```
+
+#[derive(Clone)]
+pub struct Events(Env);
+
+impl Debug for Events {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "Events")
+    }
+}
+
+pub trait Topics: IntoVal<Env, Vec<RawVal>> {}
+
+impl IntoVal<Env, Vec<RawVal>> for () {
```

Comment:
Todo: move this and the macro for tuples into tuple.rs or similar since it doesn't only apply to events but more generally is for supporting zero length tuple to Vec conversions.

Reply:
I don't think tuple.rs is the right place since `Vec<T>` is defined in the sdk, so maybe somewhere in the sdk (`vec.rs`?).

Reply:
I was meaning in tuple.rs in the SDK, a new file. Inside vec.rs is okay too, that file is just getting big.

Reply:
Ah, Ok. Yeah that makes sense.